contracts/libraries/UniswapRouter.sol

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.4;
+
+import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {ISwapRouter} from "../interfaces/ISwapRouter.sol";
+import "./Path.sol";
+
+library UniswapRouter {
+    using Path for bytes;
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    struct SwapParams {
+        bytes path;
+        address recipient;
+        address tokenIn;
+        address tokenOut;
+        uint256 amountIn;
+        uint256 minAmountOut;
+        address router;
+    }
+
+    event Swap(
+        address indexed tokenIn,
+        address indexed tokenOut,
+        uint256 amountIn,
+        uint256 amountOut
+    );
+
+    function decodePath(bytes memory path)
+        internal
+        pure
+        returns (address tokenIn, address tokenOut)
+    {
+        uint24 fee;
+
+        (tokenIn, tokenOut, fee) = path.decodeFirstPool();
+        bool hasMultiplePools = path.hasMultiplePools();
+
+        if (hasMultiplePools) {
+            path = path.skipToken();
+            while (true) {
+                (, tokenOut, ) = path.decodeFirstPool();
+
+                hasMultiplePools = path.hasMultiplePools();
+
+                if (hasMultiplePools) {
+                    path = path.skipToken();
+                } else {
+                    break;
+                }
+            }
+        }
+
+        return (tokenIn, tokenOut);
+        // return (address(0), address(0));
+    }
+
+    function getTokenOut(bytes memory path)
```

Comment:
Same here, can use calldata https://github.com/ribbon-finance/ribbon-v2/pull/167/files#r756666724

Reply:
This function is redundant, will remove.

Reply:
Addressed [here](https://github.com/ribbon-finance/ribbon-v2/pull/167#discussion_r757028909).